cdc/metadata/metadata.go

```diff
@@ -0,0 +1,232 @@
+
+// SchedState is the type of state to schedule owners and processors.
+type SchedState int
+
+const (
+	// SchedLaunched means the owner or processor is launched.
+	SchedLaunched SchedState = iota
+	// SchedRemoving means the owner or processor is in removing.
+	SchedRemoving
+	// SchedRemoved means the owner or processor is removed.
+	SchedRemoved
+)
+
+// StateToString convert state to string.
+func StateToString(state SchedState) string {
+	switch state {
+	case SchedLaunched:
+		return "Launched"
+	case SchedRemoving:
+		return "Removing"
+	case SchedRemoved:
+		return "Removed"
+	}
+	return "unreachable"
+}
+
+// ScheduledOwner is for owner and processor schedule.
+type ScheduledOwner struct {
+	*ChangefeedInfo
+	ChangefeedProgress
+	Capture *CaptureInfo
+	State   SchedState
+}
+
+// ScheduledProcessor is for owner and processor schedule.
+type ScheduledProcessor struct {
+	*ChangefeedInfo
+	ChangefeedProgress
+	Capture *CaptureInfo
+	State   SchedState
+}
+
+// ChangefeedSchedule is used to query changefeed schedule information.
+type ChangefeedSchedule struct {
+	Owner     ScheduledOwner
+	Processor []ScheduledProcessor
+}
+
+// CaptureObservation is for observing and updating metadata on a CAPTURE instance.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type CaptureObservation interface {
+	// CaptureInfo tells the caller who am I.
+	CaptureInfo() *CaptureInfo
+
+	// GetChangefeed queries one changefeed or all changefeeds.
+	GetChangefeeds(...model.ChangeFeedID) ([]*ChangefeedInfo, error)
+
+	// GetCaptures queries one capture or all captures.
+	GetCaptures(...string) ([]*CaptureInfo, error)
+
+	// Heartbeat tells the metadata storage I'm still alive.
+	Heartbeat(context.Context) error
+
+	// TakeControl blocks until becomes controller or gets canceled.
+	TakeControl() (ControllerObservation, error)
+
+	// Advance advances some changefeed progresses.
+	Advance(cfs []*ChangefeedInfo, progresses []*ChangefeedProgress) error
+
+	// Fetch the latest changefeed owner list from metadata.
+	RefreshOwners() <-chan []ScheduledOwner
+
+	// When an owner exits, inform the metadata storage.
+	PostOwnerRemoved(cf *ChangefeedInfo) error
+
+	// Fetch the latest changefeed processors from metadata.
+	RefreshProcessors() <-chan []ScheduledProcessor
+
+	// When a processor exits, inform the metadata storage.
+	PostProcessorRemoved(cf *ChangefeedInfo) error
+}
+
+// ControllerObservation is for observing and updating meta by Controller.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type ControllerObservation interface {
+	// CreateChangefeed creates a changefeed.
+	CreateChangefeed(cf *ChangefeedInfo, up *UpstreamInfo) error
+
+	// RemoveChangefeed removes a changefeed, will auto stop owner and processors.
+	RemoveChangefeed(cf *ChangefeedInfo) error
+
+	// Fetch the latest capture list in the TiCDC cluster.
+	RefreshCaptures() <-chan []*CaptureInfo
```

Comment:
If CaptureInfo  returned in GetChangefeedSchedule, RefreshCaptures may not be required?